### PR TITLE
Add initial Puppet integration RFC

### PR DESCRIPTION
**Pull request by impl**
[Rendered](https://github.com/puppetlabs/relay-rfcs/blob/puppet-integration/content/0000-puppet-integration/rfc.md)

### CODE REVIEW

**Comment by kenazk** [comment]
🔥🔥🔥

**Review by kyleterry — APPROVED** [review]
Thanks for noting the clunkiness. I'm good with this approach for now.

File: content/0000-puppet-integration/rfc.md [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+the Relay SaaS service, we'll use a Puppet module that includes plugins to send
+data to workflows via triggers (outbound) and a scheduled process that polls for
+requests to run Puppet on specific nodes (inbound). In particular, we want to
+avoid complex or undesirable networking scenarios like maintaining an
+out-of-band VPN connection or opening ports on corporate firewalls.
+
+### Module structure
+
+Our Puppet module defines one plugin, a report processor to run Relay workflows,
+and one class, `relay::bridge`, that installs the report processor (using
```
Comment (hunner) [inline content/0000-puppet-integration/rfc.md]: I called it `relay::reporting` for now but this is a better name.

File: content/0000-puppet-integration/rfc.md [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+[`ini_setting`](https://github.com/puppetlabs/puppetlabs-inifile) or
+`pe_ini_setting`) and a bidirectional process to handle requests to run Puppet.
+
+### Report processor
+
+The report processor sends report data to one or more workflow push triggers.
```
Comment (hunner) [inline content/0000-puppet-integration/rfc.md]: Currently the push triggers are 1:1 token:workflow, and the report processor just takes a single token. I can update this to take a list of tokens, or just wait until the puppet connection is ready.

File: content/0000-puppet-integration/rfc.md [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+1. Using the configured execution mechanism, either for PE or open source
+   Puppet, starts the Puppet run.
+1. Updates the state of the run to `in-progress` in the Relay SaaS API.
+1. Continues to monitor the run's progress. As a heartbeat mechanism, updates
+   the state of the run in the Relay API periodically.
+1. When the run completes, whether successful or not, updates the state of the
+   run in the Relay API to indicate the completion.
+1. Exits.
+
+We schedule the bridge process using a systemd timer that runs on an interval
```
Comment (hunner) [inline content/0000-puppet-integration/rfc.md]: All enterprise puppetserver platforms use systemd, so this is fine.

File: content/0000-puppet-integration/rfc.md [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+
+  ```yaml
+  connection: !Connection {type: puppet, name: my-puppet-connection}
+  timeoutAfterMissedHeartbeats: 3
+  environment: staging
+  scope:
+    nodes:
+    - node-a.example.com
+    - node-b.example.com
+  noOp: false
```
Comment (hunner) [inline content/0000-puppet-integration/rfc.md]: 🤔

File: content/0000-puppet-integration/rfc.md [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+
+As stated, this is intended to be alpha-quality software. We have a slight risk
+of not adequately communicating this and being forced to maintain the APIs
+defined here beyond their usable lifetimes.
+
+While this design does not require firewall changes for most on-premise networks
+that allow outbound internet traffic, it cannot work in a fully air-gapped
+environment. It also exposes internal PE node data to our SaaS service if the
+report processor plugin is used. Some customers may not be able to use this
+module for regulatory or compliance reasons.
```
Comment (hunner) [inline content/0000-puppet-integration/rfc.md]: Especially considering sensitive information in the logs as in the gotcha section of https://puppet.com/blog/encrypt-your-data-using-hiera-eyaml/

**Review by hunner — APPROVED** [review]
Nice and thorough :)

**Review by grimradical — APPROVED** [review]

**Pull request merged**